datafusion/physical-expr/src/expressions/if_fun.rs

```diff
@@ -0,0 +1,175 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::array::{Array, ArrayRef};
+use arrow::compute::kernels::zip::zip;
+use datafusion_common::{cast::as_boolean_array, DataFusionError, Result, ScalarValue};
+use datafusion_expr::ColumnarValue;
+
+/// Implements IF(expr1, expr2, expr3)
+/// If `expr1` evaluates to true, then returns `expr2`; otherwise returns `expr3`.
+/// Args: 0 - boolean array
+///       1 - if expr1 is true, then the result is expr2
+///       2 - if expr1 is true, then the result is expr3.
+///
+pub fn if_func(args: &[ColumnarValue]) -> Result<ColumnarValue> {
+    if args.len() != 3 {
+        return Err(DataFusionError::Internal(format!(
+            "{:?} args were supplied but IF takes exactly 3 args",
+            args.len(),
+        )));
+    }
+    let mut array_lens = args.iter().filter_map(|x| match x {
+        ColumnarValue::Array(array) => Some(array.len()),
+        _ => None,
+    });
+
+    let (predicates, true_vals, false_vals) = (&args[0], &args[1], &args[2]);
+
+    if let Some(size) = array_lens.next() {
+        match predicates {
+            ColumnarValue::Scalar(_) => {
+                handle_scalar_predicates(predicates, true_vals, false_vals)
+            }
+            ColumnarValue::Array(predicates) => {
+                handle_array_predicates(predicates, size, true_vals, false_vals)
+            }
+        }
+    } else {
+        handle_scalar_predicates(predicates, true_vals, false_vals)
+    }
+}
+
+fn handle_scalar_predicates(
+    predicates: &ColumnarValue,
+    true_vals: &ColumnarValue,
+    false_vals: &ColumnarValue,
+) -> Result<ColumnarValue> {
+    match predicates {
+        ColumnarValue::Scalar(ScalarValue::Boolean(Some(predicates))) => {
+            if *predicates {
+                Ok(true_vals.clone())
+            } else {
+                Ok(false_vals.clone())
+            }
+        }
+        ColumnarValue::Scalar(val) if val.is_null() => Ok(false_vals.clone()),
```

Comment:
Shouldn't we check the datatype of `val`?

Reply:
Hi @viirya ,here I just want to handle when the predicate expression is constant null,  then return the false values, like function: if(null, 1, 2) will return 2.
null values for any data type can be considered as false, and I just add is_null check for the val, I think no need to check the datatype